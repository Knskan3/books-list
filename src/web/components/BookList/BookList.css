.bookListContainer {
  margin-left: 20px;
  width: 261px;
  height: 455px;
  border: 1px solid #ccc;
  overflow: auto;
}

.bookRow {
  height: 100px;
  border-bottom: 1px solid #cccccc;
  padding: 10px;
}

.loadingImage {
  display: block;
  width: 100px;
  margin: 0 auto;
}

.hide {
  display: none;
}

.wrapper {
  width: 301px;
  height: 636px;
  padding-top: 80px;
  background: url('../../styles/img/phone.png') no-repeat;
  background-size: 100%;
  box-sizing: border-box;
}

.genderIcon {
  height: 17px;
  width: 17px;
  margin-right: 5px;
}

.label {
  margin-right: 5px;
}
