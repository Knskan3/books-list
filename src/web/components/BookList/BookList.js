import React, { Component } from 'react';
import PropTypes from 'prop-types';
import {
  Label,
} from 'react-bootstrap';
import genders from '../../../book-generator/genders';
import styles from './BookList.css';
import loadingImage from '../../styles/img/loading.gif';
import maleIcon from '../../styles/img/male.png';
import femaleIcon from '../../styles/img/female.png';

/**
 * Describes the BookList react component
 */
class BookList extends Component {
  /**
   * Creates a BookList
   *
   * @param {Object} props React component props injected by connect and passed as attributes
   *
   * @public
   */
  constructor(props) {
    super(props);
    this.bottomTreshold = 70;
    this.topTreshold = 30;
    this.props.bookRequest({ reset: true });
  }

  /**
   * Lifecycle component method.
   * We use it to add a event listener and handler to the scroll event.
   */
  componentDidMount() {
    this.scrollerElement.addEventListener('scroll', this.scrollHandler);
  }

  /**
   * Lifecycle component method.
   * We use it to cheat the top appending not updating scroll
   */
  componentDidUpdate() {
    if (this.scrollerElement.scrollTop === 0) {
      this.scrollerElement.scrollTop = 1;
    }
  }

  /**
   * Formats the book date in a better readable format
   * @param {Object} book
   * @returns {string} Format date
   */
  getFormatedDate = (book) => {
    const d = new Date(book.publishDate);
    return `${d.getDate()}/${(d.getMonth() + 1)}/${d.getFullYear()}`;
  }

  /**
   * Generates a component with the author gender icon
   *
   * @param gender
   * @returns {JSX} Image component with gender icon
   */
  getAuthorGenderIcon = (gender) => {
    if (gender === genders.FEMALE) {
      return <img className={styles.genderIcon} src={femaleIcon} alt="female" />;
    }
    return <img className={styles.genderIcon} src={maleIcon} alt="male" />;
  }

  /**
   * Stores the DOM node containing the scroll
   *
   * @param {DOMElement}
   */
  captureScroller = (node) => {
    this.scrollerElement = node;
  }

  /**
   * Handles the scroll event. The idea is to request books if the scroll is situated between
   * some thresholds. Either going up or down.
   *
   * @param {Event} e Event containing scroll info
   */
  scrollHandler = (e) => {
    const node = e.srcElement;
    const scrollPer = (node.scrollTop / (node.scrollHeight - node.offsetHeight)) * 100;

    // scroll down
    if ((!this.props.books.loading) &&
      (scrollPer > this.bottomTreshold) &&
      (this.props.books.total > this.props.books.last)) {
      this.props.bookRequest({ reset: false });
    }

    // scroll up
    if ((!this.props.books.loadingBackwards) &&
      (scrollPer < this.topTreshold) &&
      (this.props.books.first > 0)) {
      this.props.bookRequest({ reset: false, backwards: true });
    }
  }

  /**
   * Lifecycle component method. Renders the component.
   *
   * @returns {JSX}
   */
  render() {
    return (<div className={styles.wrapper}>
      <div id="scroller" ref={this.captureScroller} className={styles.bookListContainer}>
        <div className={this.props.books.loadingBackwards ? '' : styles.hide}>
          <div><img className={styles.loadingImage} src={loadingImage} alt="loading" /></div>
        </div>
        {
          this.props.books.stack.map((book) => (
            // eslint-disable-next-line
            <div className={`bookItem ${styles.bookRow}`} key={book._id}>
              <div><b>{book.name}</b></div>
              <div>By: {book.author.name}</div>
              <div>On: {this.getFormatedDate(book)}</div>
              <div>
                {this.getAuthorGenderIcon(book.author.gender)}
                <Label className={styles.label} bsStyle="primary">{book.genre.name}</Label>
                {(book.horrorFlag)
                  && (<Label className={`superHorror ${styles.label}`} bsStyle="danger">Super Horror</Label>)}
                {(book.financeFlag)
                && (<Label className={`hotFinance ${styles.label}`} bsStyle="info">Hot Finance</Label>)}
              </div>
            </div>
          ))
        }
        <div className={this.props.books.loading ? '' : styles.hide}>
          <div><img className={styles.loadingImage} src={loadingImage} alt="loading" /></div>
        </div>
        <div className={(!this.props.books.loading && (this.props.books.stack.length <= 0)) ? '' : styles.hide}>
          <div className={styles.bookRow} >No Results :(</div>
        </div>
      </div>
    </div>);
  }
}

/**
 * Props type validation for this component.
 * @type {{books: *, bookRequest: *}}
 */
BookList.propTypes = {
  books: PropTypes.object,
  bookRequest: PropTypes.func,
};

export default BookList;
