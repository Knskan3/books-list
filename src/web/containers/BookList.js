/**
 * Created by javier on 28/09/2017.
 */
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import BookListComponent from '../components/BookList/BookList';
import {
  booksStateSelector,
} from '../selectors';
import { bookRequest } from '../actions';

/**
 * Injects the needed state values into the component as props.
 * We are using selectors from reselect to memoize and improve performance
 * This will also allow us to access state easily form sagas.
 *
 * @param {Object} state Cureent app state
 */
const mapStateToProps = (state) => ({
  books: booksStateSelector(state),
});

/**
 * Injects actions into the component class props
 *
 * @param {Function} dispatch Action dispatcher
 * @returns {Function} dispatchProps
 */
const mapDispatchToProps = (dispatch) => (
  bindActionCreators({
    bookRequest,
  }, dispatch)
);

export default connect(mapStateToProps, mapDispatchToProps)(BookListComponent);

