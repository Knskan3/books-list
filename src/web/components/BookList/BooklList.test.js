import React from 'react';
import Enzyme, { mount } from 'enzyme';
import renderer from 'react-test-renderer';
import Adapter from 'enzyme-adapter-react-15';
import BookList from './BookList';

Enzyme.configure({ adapter: new Adapter() });

const scrollEventListenerSpy = jest.fn();

/**
 * This method will be injected in the renderer in order to
 * mock the ref in the component. Those renderers doesnt really access the DOM
 * so we have to mock the responses.
 *
 * @param {Object} element Object generated by react
 *
 * @return {Object}
 */
const createNodeMock = (element) => {
  if (element.props.id === 'scroller') {
    return {
      addEventListener: scrollEventListenerSpy,
    };
  }
  return {};
};

const defaultBookState = {
  stack: [],
  first: 0,
  last: 0,
  total: 0, // in db without filter
  loading: false,
  loadingBackwards: false,
  batchSize: 150,
};
/* eslint-disable */

const bookMocks = [{
  '_id': '59ceadc9485517c5b4b71d88',
  'name': 'Ben uniip.',
  'author': {'name': 'Randee Filberte', 'gender': 'MALE'},
  'genre': {'id': 23, 'name': 'Series'},
  'publishDate': '1929-05-04T17:39:07.390Z',
  'index': 6,
  'horrorFlag': false,
  'financeFlag': false,
}, {
  '_id': '59ceadc9485517c5b4b71d8a',
  'name': 'Ofe mevigi.',
  'author': {'name': 'Beckie Llywellyn', 'gender': 'FEMALE'},
  'genre': {'id': 4, 'name': 'Drame'},
  'publishDate': '1938-08-19T13:38:30.424Z',
  'index': 8,
  'horrorFlag': false,
  'financeFlag': false,
}];
/* eslint-enable */

describe('<BookList />', () => {
  // Snapshot testing
  test('Snapshot sould be correct', () => {
    const bookRequestSpy = jest.fn();
    const component = renderer.create(
      <BookList books={defaultBookState} bookRequest={bookRequestSpy} />,
      { createNodeMock },
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  test('Should request books on constructor', () => {
    const bookRequestSpy = jest.fn();
    mount(<BookList books={defaultBookState} bookRequest={bookRequestSpy} />);
    expect(bookRequestSpy.mock.calls.length).toBe(1);
  });

  test('Should display correct number of books', () => {
    const bookRequestSpy = jest.fn();
    const books = {
      ...defaultBookState,
      stack: bookMocks,
    };
    const component = mount(<BookList books={books} bookRequest={bookRequestSpy} />);
    expect(component.find('#scroller div.bookItem').length).toBe(2);
  });

  test('Should not show special flags if none', () => {
    const bookRequestSpy = jest.fn();
    const books = {
      ...defaultBookState,
      stack: [{
        ...bookMocks[0],
        horrorFlag: false,
        financeFlag: false,
      }],
    };
    const component = mount(<BookList books={books} bookRequest={bookRequestSpy} />);
    expect(component.find('div.bookItem .superHorror').length).toBe(0);
    expect(component.find('div.bookItem .hotFinance').length).toBe(0);
  });

  test('Should show special flags if available', () => {
    const bookRequestSpy = jest.fn();
    const books = {
      ...defaultBookState,
      stack: [{
        ...bookMocks[0],
        horrorFlag: true,
        financeFlag: true,
      }],
    };
    const component = mount(<BookList books={books} bookRequest={bookRequestSpy} />);
    expect(component.find('div.bookItem span.superHorror').length).toBe(1);
    expect(component.find('div.bookItem span.hotFinance').length).toBe(1);
  });

  test('Should generate proper gender icon (male)', () => {
    const bookRequestSpy = jest.fn();
    const books = {
      ...defaultBookState,
      stack: [{
        ...bookMocks[0],
      }],
    };
    const component = mount(<BookList books={books} bookRequest={bookRequestSpy} />);
    expect(component.find('img[alt="male"]').length).toBe(1);
  });

  test('Should generate proper gender icon (female)', () => {
    const bookRequestSpy = jest.fn();
    const books = {
      ...defaultBookState,
      stack: [{
        ...bookMocks[1],
      }],
    };
    const component = mount(<BookList books={books} bookRequest={bookRequestSpy} />);
    expect(component.find('img[alt="female"]').length).toBe(1);
  });

  describe('Scroll Handler', () => {
    test('Should not request more books up if loading', () => {
      const bookRequestSpy = jest.fn();
      // scroll up but loading
      const elementMock = {
        srcElement: {
          scrollTop: 0,
          scrollHeight: 1000,
          offsetHeight: 500,
        },
      };
      const books = {
        ...defaultBookState,
        loading: true,
        loadingBackwards: true,
        total: 1000000,
        last: 1000,
        first: 500,
      };
      const component = mount(<BookList books={books} bookRequest={bookRequestSpy} />);
      component.instance().scrollHandler(elementMock);
      expect(bookRequestSpy.mock.calls.length).toBe(1);
    });

    test('Should not request more books down if loading', () => {
      const bookRequestSpy = jest.fn();
      // scroll up but loading
      const elementMock = {
        srcElement: {
          scrollTop: 1000,
          scrollHeight: 1000,
          offsetHeight: 500,
        },
      };
      const books = {
        ...defaultBookState,
        loading: true,
        loadingBackwards: true,
        total: 1000000,
        last: 1000,
        first: 500,
      };
      const component = mount(<BookList books={books} bookRequest={bookRequestSpy} />);
      component.instance().scrollHandler(elementMock);
      expect(bookRequestSpy.mock.calls.length).toBe(1);
    });

    test('Should request more books down if scroll down and not loading', () => {
      const bookRequestSpy = jest.fn();
      // scroll up but loading
      const elementMock = {
        srcElement: {
          scrollTop: 990,
          scrollHeight: 1000,
          offsetHeight: 500,
        },
      };
      const books = {
        ...defaultBookState,
        loading: false,
        loadingBackwards: true,
        total: 1000000,
        last: 1000,
        first: 500,
      };
      const component = mount(<BookList books={books} bookRequest={bookRequestSpy} />);
      component.instance().scrollHandler(elementMock);
      expect(bookRequestSpy.mock.calls.length).toBe(2);
    });

    test('Should request more books up if scroll up and not loading', () => {
      const bookRequestSpy = jest.fn();
      // scroll up but loading
      const elementMock = {
        srcElement: {
          scrollTop: 0,
          scrollHeight: 1000,
          offsetHeight: 500,
        },
      };
      const books = {
        ...defaultBookState,
        loading: true,
        loadingBackwards: false,
        total: 1000000,
        last: 1000,
        first: 500,
      };
      const component = mount(<BookList books={books} bookRequest={bookRequestSpy} />);
      component.instance().scrollHandler(elementMock);
      expect(bookRequestSpy.mock.calls.length).toBe(2);
    });
  });
});
