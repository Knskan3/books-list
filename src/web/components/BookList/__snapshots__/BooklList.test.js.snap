// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BookList /> Snapshot sould be correct 1`] = `
<div
  className={undefined}
>
  <div
    className={undefined}
    id="scroller"
  >
    <div
      className={undefined}
    >
      <div>
        <img
          alt="loading"
          className={undefined}
          src={Object {}}
        />
      </div>
    </div>
    <div
      className={undefined}
    >
      <div>
        <img
          alt="loading"
          className={undefined}
          src={Object {}}
        />
      </div>
    </div>
    <div
      className=""
    >
      <div
        className={undefined}
      >
        No Results :(
      </div>
    </div>
  </div>
</div>
`;
